src/main/java/servlets/SearchServlet.java

```diff
@@ -38,33 +39,102 @@
 /** Servlet that searches and returns matching receipts from datastore. */
 @WebServlet("/search-receipts")
 public class SearchServlet extends HttpServlet {
+  /** Messages that show up on client-side banner on thrown exception. */
+  private static final String NULL_EXCEPTION_MESSAGE =
+      "Null Field: Receipt unable to be queried at this time, please try again.";
+  private static final String NUMBER_EXCEPTION_MESSAGE =
+      "Invalid Price: Receipt unable to be queried at this time, please try again.";
+  private static final String PARSE_EXCEPTION_MESSAGE =
+      "Dates Unparseable: Receipt unable to be queried at this time, please try again.";
+
+  private final DatastoreService datastore;
+
+  public SearchServlet() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  public SearchServlet(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String categories = request.getParameter("categories");
-    ImmutableList<Receipt> receipts = getReceiptsWithMatchingCategories(categories);
+    QueryInformation queryInformation = createQueryInformation(request, response);
+
+    if (queryInformation == null) { // createQueryInformation() threw an exception, so exit
+      return;
+    }
+
+    ImmutableList<Receipt> receipts = getMatchingReceipts(queryInformation);
 
     Gson gson = new Gson();
     response.setContentType("application/json;");
     response.getWriter().println(gson.toJson(receipts));
   }
 
-  /** Returns ImmutableList of receipts from datastore with the same categories. */
-  private ImmutableList<Receipt> getReceiptsWithMatchingCategories(String categories) {
-    // Set filter to retrieve only receipts with categories equal to categories.
-    Filter matchingCategories = new FilterPredicate("categories", FilterOperator.EQUAL, categories);
-    Query query = new Query("Receipt");
-    query.setFilter(matchingCategories);
+  /** Creates a {@link QueryInformation} based on request parameters. */
+  private QueryInformation createQueryInformation(
+      HttpServletRequest request, HttpServletResponse response) throws IOException {
```

Comment:
Not mandatory but this would be a good place to use the Optional container type as a return type.

https://www.oracle.com/technical-resources/articles/java/java8-optional.html
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html

Reply:
This looks helpful! I'll try to get back to it if I have time